File: C++/섹션05. 완전 탐색 기본/02-02. BOJ 2606 - 바이러스 (DFS, 인접 리스트).cpp
#include <stdio.h>

#define MAX (100 + 10)

int V, E; // Vertex, Edge
int MAP[MAX][MAX];
int index[MAX];

bool visit[MAX];
int count;

void input()
{
	scanf("%d %d", &V, &E);

	for (int i = 0; i < E; i++)
	{
		int n1, n2;

		scanf("%d %d", &n1, &n2);

		MAP[n1][index[n1]++] = n2;
		MAP[n2][index[n2]++] = n1;
	}
}

void printMap() // for debug
{
	for (int r = 1; r <= V; r++)
	{
		for (int c = 0; c < index[r]; c++)
			printf("%d ", MAP[r][c]);
		putchar('\n');
	}
	putchar('\n');
}

void DFS(int node)
{
	visit[node] = true;

	for (int i = 0; i < index[node]; i++)
	{
		int number = MAP[node][i];

		if (visit[number] == true) continue;

		count++;

		DFS(number);
	}
}

int main()
{
	input();

	// printMap();

	DFS(1);

	printf("%d\n", count);

	return 0;
}

File: Python/섹션05. 완전 탐색 기본/02-02. BOJ 2606 - 바이러스 (DFS, 인접 리스트).py
MAX = 100 + 10

MAP = [[0] * MAX for _ in range(MAX)]
index = [0] * MAX
visit = [False] * MAX

count = 0

def input_data():
    global V, E

    V = int(input()) # Vertex
    E = int(input()) # Edge

    for _ in range(E):
        n1, n2 = map(int, input().split())
        
        MAP[n1][index[n1]] = n2
        index[n1] += 1
        MAP[n2][index[n2]] = n1
        index[n2] += 1
        
def print_map(): # for debug
    for r in range(1, V + 1):
        print(*MAP[r][:index[r]])
    print("")

def DFS(node):      
    global count
      
    visit[node] = True

    for i in range(index[node]):        
        number = MAP[node][i]
        
        if visit[number] == True: continue

        count = count + 1

        DFS(number)
       
input_data()
        
DFS(1)

print(count)
